### PR TITLE
_dependency_source, _cli, test: Do not override `pip.conf` unless explicitly specified via flags

### LINKED ISSUE #563: Private repo dependencies no longer being skipped
## Bug description

Private repo dependencies are no longer being skipped if they cannot be resolved.
For example, in the v2.4.14, the audit would succeed and then show that the packages were skipped:
image

But now in v2.5.x, the audit fails and shows the error:
image

## Reproduction steps

- Ensure `pip-audit` v2.5.x is installed
- Create a `requirements.txt` file with a dependency that is installed from a private repo
- Run `pip-audit -r requirements.txt`

## Expected behavior

I'd expect the same behaviour as v2.4.14 - whereby the private dependencies would simply be skipped - unless the `-S` flag was specified.

## Platform information

* OS name and version:
* `pip-audit` version (`pip-audit -V`): 2.5.2
* Python version (`python -V` or `python3 -V`): 3.11.2
* `pip` version (`pip -V` or `pip3 -V`): 23.0.1

### PR DESCRIPTION
**Pull request by philblckwd**
Refactor index-url and extra-index-url options to not override config defined in 'pip.conf' unless explicitly specified via flags.

Closes #563

### CODE REVIEW

File: pip_audit/_cli.py
Comment (woodruffw) [inline pip_audit/_cli.py:289]: I think we want to keep this `default` in the CLI -- behaviorally it should remain the default, and having it appear as such in the `--help` render makes for a nicer UX 🙂

File: pip_audit/_dependency_source/pyproject.py
Comment (woodruffw) [inline pip_audit/_dependency_source/pyproject.py:33]: Similar typestate nit here: maybe drop the default `""` here (or make it `PYPI_URL`), since we always expect a valid URL?

File: pip_audit/_dependency_source/requirement.py
Comment (woodruffw) [inline pip_audit/_dependency_source/requirement.py:49]: Same as https://github.com/pypa/pip-audit/pull/565/files#r1145039056

Comment (woodruffw) [reply pip_audit/_dependency_source/requirement.py:49]:
```suggestion
        index_url: str | None = None,
```

**Comment by woodruffw** [comment]
Thanks a ton for opening this, @philblckwd! 

This looks great to me structurally; just a few small nitpicks 🙂

**Comment by woodruffw** [comment]
(This also needs a `CHANGELOG.md entry. Feel free to do it yourself if you feel like it, otherwise I'll do it as a fixup after merge!)

File: pip_audit/_cli.py
Comment (philblckwd) [reply pip_audit/_cli.py:289]: So leaving this as the default means that it will always be explicitly passed to the `pip install` CLI, which results in the `index-url` specified in the `pip.conf` being ignored. I understand that it makes for a more friendly UX to specify as such, but would it be reasonable to convey that the url will be resolved by `pip` if not passed by the user explicitly?

If this is a deal breaker though, then I'll have to dig deeper into `pip` to determine how it is performing the `index-url` resolution.

File: pip_audit/_dependency_source/pyproject.py
Comment (tetsuo-cpp) [reply pip_audit/_dependency_source/pyproject.py:33]: Yep, I'd prefer this too.

**Comment by tetsuo-cpp** [comment]
I'm also happy with where this is at. Once the help text is amended and those type state related comments are addressed, this should be good to go.

Thanks again for your work. We really appreciate it.

**Comment by tetsuo-cpp** [comment]
@philblckwd One more thing. We have a check in CI to ensure that the help text that we display in the README is up to date. So once you've amended the help text for `--index-url`, you'll need to update the README.

**Comment by philblckwd** [comment]
Per feedback:

- I updated arguments to use optional types
- I updated the help text, as well as the readme, to state the `index-url` will be resolved by pip if not specified
- I added an entry to the changelog
- I noticed unrelated lint and check-readme errors, so resolved those too

**Review by tetsuo-cpp — APPROVED** [review]
LGTM!

**Comment by tetsuo-cpp** [comment]
I'll merge this shortly after I've done a bit of testing locally.

**Comment by tetsuo-cpp** [comment]
Thanks @philblckwd!

**Pull request merged**
